File: SatinalmaPro.Shared/Helpers/OnaylananMalzemeOlusturucu.cs
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Helpers;

/// <summary>Onaylı taleplerden mal kabul kuyruğu satırları üretir — teklifli ve teklifsiz onay.</summary>
public static class OnaylananMalzemeOlusturucu
{
    public static bool MalKabulTalep(SatinalmaTalep talep) =>
        talep.Durum is SatinalmaTalepDurumlari.Onaylandi or SatinalmaTalepDurumlari.SiparisOlusturuldu
        && (talep.HerhangiKalemOnayli || talep.TeklifsizYonetimOnayi);

    public static bool MalKabulBekleyen(OnaylananMalzemeSatiri satir) =>
        !satir.SiparisTamamlandi
        && satir.Durum is SatinalmaTalepDurumlari.Onaylandi or SatinalmaTalepDurumlari.SiparisOlusturuldu;

    public static bool SevkiyatTamamlanabilir(OnaylananMalzemeSatiri satir) =>
        !satir.SiparisTamamlandi
        && satir.KabulEdilenMiktar > 0.0001
        && satir.KabulEdilenMiktar < satir.SiparisMiktari - 0.0001
        && satir.Durum is SatinalmaTalepDurumlari.Onaylandi or SatinalmaTalepDurumlari.SiparisOlusturuldu;

    public static List<OnaylananMalzemeSatiri> Olustur(IEnumerable<SatinalmaTalep> talepler)
    {
        var liste = new List<OnaylananMalzemeSatiri>();

        foreach (var talep in talepler.Where(MalKabulTalep))
        {
            talep.FirmaSiparisNolari ??= [];
            var teklifsiz = talep.TeklifsizYonetimOnayi && !talep.HerhangiKalemOnayli;

            foreach (var kalem in talep.Kalemler
                         .Where(k => !string.IsNullOrWhiteSpace(k.Malzeme))
                         .Where(k => teklifsiz || KalemFirmaAtamaYardimcisi.OnayliMi(k))
                         .OrderBy(k => k.SiraNo))
            {
                if (teklifsiz && !KalemFirmaAtamaYardimcisi.OnayliMi(kalem))
                {
                    var siparisNo = string.IsNullOrWhiteSpace(talep.SiparisNo) ? talep.TalepNo : talep.SiparisNo;
                    liste.Add(Satir(talep, kalem, Guid.Empty, "", "", 0, 0, 0, siparisNo,
                        kalem.Miktar, kalem.KabulEdilenMiktar, kalem.SiparisTamamlandi));
                    continue;
                }

                foreach (var atama in KalemFirmaAtamaYardimcisi.EtkinAtamalar(kalem))
                {
                    var teklif = talep.Teklifler.FirstOrDefault(t => t.Id == atama.TeklifId);
                    if (teklif is null)
                        continue;

                    teklif.FiyatlariHesapla(talep.Kalemler);
                    var fiyat = teklif.Fiyatlar.FirstOrDefault(f => f.KalemId == kalem.Id);
                    if (fiyat is null)
                        continue;

                    var siparisNo = talep.FirmaSiparisNolari.TryGetValue(teklif.Id, out var no)
                                    && !string.IsNullOrWhiteSpace(no)
                        ? no
                        : talep.SiparisNo;

                    var birim = fiyat.TlBirimFiyat(teklif.UsdKuru, teklif.EurKuru);
                    var toplam = Math.Round(birim * (decimal)atama.Miktar, 2);

                    liste.Add(Satir(talep, kalem, teklif.Id, teklif.FirmaAdi,
                        string.IsNullOrWhiteSpace(fiyat.Marka) ? teklif.Marka : fiyat.Marka,
                        birim, toplam,
                        teklif.VadeGunu, siparisNo,
                        atama.Miktar, atama.KabulEdilenMiktar, atama.SiparisTamamlandi));
                }
            }
        }

        return liste;
    }

    private static OnaylananMalzemeSatiri Satir(
        SatinalmaTalep talep,
        SatinalmaTalepKalemi kalem,
        Guid teklifId,
        string firma,
        string marka,
        decimal birimFiyat,
        decimal toplamTutar,
        int vadeGunu,
        string siparisNo,
        double siparisMiktari,
        double kabulEdilenMiktar,
        bool siparisTamamlandi) => new()
    {
        TalepId = talep.Id,
        KalemId = kalem.Id,
        TeklifId = teklifId,
        TalepNo = talep.TalepNo,
        SiparisNo = siparisNo,
        Tarih = talep.Tarih,
        Durum = talep.Durum,
        Firma = firma,
        Marka = marka,
        Malzeme = kalem.Malzeme,
        SiparisMiktari = siparisMiktari,
        KabulEdilenMiktar = kabulEdilenMiktar,
        SiparisTamamlandi = siparisTamamlandi,
        Birim = kalem.Birim,
        BirimFiyati = birimFiyat,
        ToplamTutar = toplamTutar,
        KalemAciklamasi = kalem.Aciklama,
        VadeGunu = vadeGunu
    };
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/android/core/roles/OnaylananMalzemeOlusturucu.kt
package com.satinalmapro.android.core.roles

import com.satinalmapro.android.core.model.OnaylananMalzemeSatiri
import com.satinalmapro.android.core.model.TalepItem
import com.satinalmapro.android.core.model.TalepKalem

object OnaylananMalzemeOlusturucu {

    fun malKabulTalep(talep: TalepItem): Boolean =
        talep.durum in setOf(TalepDurumlari.ONAYLANDI, TalepDurumlari.SIPARIS) &&
            (talep.herhangiKalemOnayli || talep.teklifsizYonetimOnayi)

    fun siparisVerBekleyen(satir: OnaylananMalzemeSatiri): Boolean =
        satir.durum == TalepDurumlari.ONAYLANDI &&
            !satir.siparisTamamlandi &&
            satir.kalanMiktar > 0.0001

    fun malKabulBekleyen(satir: OnaylananMalzemeSatiri): Boolean =
        !satir.siparisTamamlandi && satir.durum == TalepDurumlari.SIPARIS

    fun sevkiyatTamamlanabilir(satir: OnaylananMalzemeSatiri): Boolean =
        !satir.siparisTamamlandi &&
            satir.kabulEdilenMiktar > 0.0001 &&
            satir.kabulEdilenMiktar < satir.siparisMiktari - 0.0001 &&
            satir.durum == TalepDurumlari.SIPARIS

    fun olustur(talepler: List<TalepItem>): List<OnaylananMalzemeSatiri> {
        val liste = mutableListOf<OnaylananMalzemeSatiri>()
        for (talep in talepler.filter(::malKabulTalep)) {
            val teklifsiz = talep.teklifsizYonetimOnayi && !talep.herhangiKalemOnayli
            @Suppress("UNCHECKED_CAST", "USELESS_ELVIS")
            val kalemler = ((talep.kalemler as List<TalepKalem>?) ?: emptyList())
                .filter { it.malzeme.isNotBlank() }
                .filter { teklifsiz || KalemFirmaAtamaYardimcisi.onayliMi(it) }
                .sortedBy { it.siraNo }
            for (kalem in kalemler) {
                if (teklifsiz && !KalemFirmaAtamaYardimcisi.onayliMi(kalem)) {
                    liste += teklifsizSatir(talep, kalem)
                    continue
                }
                for (atama in KalemFirmaAtamaYardimcisi.etkinAtamalar(kalem)) {
                    liste += atamaSatiri(talep, kalem, atama.teklifId, atama.miktar, atama.kabulEdilenMiktar, atama.siparisTamamlandi)
                }
            }
        }
        return liste
    }

    private fun teklifsizSatir(talep: TalepItem, kalem: TalepKalem): OnaylananMalzemeSatiri {
        val siparisNo = talep.siparisNo.ifBlank { talep.talepNo }
        return OnaylananMalzemeSatiri(
            talepId = talep.id,
            kalemId = kalem.id,
            talepNo = talep.talepNo,
            siparisNo = siparisNo,
            tarih = talep.tarih,
            durum = talep.durum,
            malzeme = kalem.malzeme,
            siparisMiktari = kalem.miktar,
            kabulEdilenMiktar = kalem.kabulEdilenMiktar,
            siparisTamamlandi = kalem.siparisTamamlandi,
            birim = kalem.birim
        )
    }

    private fun atamaSatiri(
        talep: TalepItem,
        kalem: TalepKalem,
        teklifId: String,
        siparisMiktari: Double,
        kabulEdilenMiktar: Double,
        siparisTamamlandi: Boolean
    ): OnaylananMalzemeSatiri {
        @Suppress("UNCHECKED_CAST", "USELESS_ELVIS")
        val teklifler = (talep.teklifler as List<com.satinalmapro.android.core.model.TeklifItem>?) ?: emptyList()
        val teklif = teklifler.firstOrNull { it.id.equals(teklifId, true) }
        @Suppress("UNCHECKED_CAST", "USELESS_ELVIS")
        val fiyatlar = (teklif?.fiyatlar as List<com.satinalmapro.android.core.model.TeklifFiyat>?) ?: emptyList()
        val fiyat = fiyatlar.firstOrNull { it.kalemId == kalem.id }
        @Suppress("UNCHECKED_CAST", "USELESS_ELVIS")
        val siparisMap = (talep.firmaSiparisNolari as Map<String, String>?) ?: emptyMap()
        val siparisNo = siparisMap[teklifId] ?: talep.siparisNo.ifBlank { talep.talepNo }
        val birimFiyat = fiyat?.birimFiyat ?: 0.0
        return OnaylananMalzemeSatiri(
            talepId = talep.id,
            kalemId = kalem.id,
            teklifId = teklifId,
            talepNo = talep.talepNo,
            siparisNo = siparisNo,
            tarih = talep.tarih,
            durum = talep.durum,
            firma = teklif?.firmaAdi.orEmpty(),
            marka = fiyat?.marka?.ifBlank { teklif?.marka.orEmpty() }.orEmpty(),
            malzeme = kalem.malzeme,
            siparisMiktari = siparisMiktari,
            kabulEdilenMiktar = kabulEdilenMiktar,
            siparisTamamlandi = siparisTamamlandi,
            birim = kalem.birim,
            birimFiyati = birimFiyat
        )
    }
}
